File: JavaScript/6 kyu Duplicate Encoder/duplicateEncode.js
/*
Write a function that takes a string and returns an encoded version of it.

Each character's first occurrence should be replaced with "(" and 
each character's subsequent occurrences should be replaced with ")".

The encoding is case-insensitive.

Example:
duplicateEncode("din"); // expected result: "((("
duplicateEncode("recede"); // expected result: "()()()"
duplicateEncode("Success"); // expected result: ")())())"
duplicateEncode("(( @"); // expected result: "))(("

Notes:
- The input string will only contain characters from the ASCII table.
- The string can be empty.
*/

// Solution

function duplicateEncode(word) {
  const lower = word.toLowerCase();
  const freq = {};

  for (const char of lower) {
    freq[char] = (freq[char] || 0) + 1;
  }

  return lower
    .split("")
    .map((char) => (freq[char] > 1 ? ")" : "("))
    .join("");
}

File: Rust/6 kyu Duplicate Encoder/DuplicateEncode.rs
// The goal of this exercise is to convert a string to a new string where each character
// in the new string is "(" if that character appears only once in the original string,
// or ")" if that character appears more than once in the original string.
// Ignore capitalization when determining if a character is a duplicate.

// Examples:
// "din"      =>  "((("
// "recede"   =>  "()()()"
// "Success"  =>  ")())())"
// "(( @"     =>  "))(("

// Solution

use std::collections::HashMap;

fn duplicate_encode(word: &str) -> String {
    let lower: String = word.chars().map(|c| c.to_lowercase().next().unwrap()).collect();
    let mut counts = HashMap::new();
    for c in lower.chars() {
        *counts.entry(c).or_insert(0) += 1;
    }
    lower.chars()
        .map(|c| if counts[&c] > 1 { ')' } else { '(' })
        .collect()
}
